build.rs

```diff
@@ -76,3 +77,26 @@ fn commit_info() {
     println!("cargo:rustc-env=CARGO_COMMIT_SHORT_HASH={}", next());
     println!("cargo:rustc-env=CARGO_COMMIT_DATE={}", next())
 }
+
+#[allow(clippy::disallowed_methods)]
+fn windows_manifest() {
+    use std::env;
+    let target_os = env::var("CARGO_CFG_TARGET_OS");
+    let target_env = env::var("CARGO_CFG_TARGET_ENV");
+    if Ok("windows") == target_os.as_deref() && Ok("msvc") == target_env.as_deref() {
+        static WINDOWS_MANIFEST_FILE: &str = "windows.manifest.xml";
+
+        let mut manifest = env::current_dir().unwrap();
+        manifest.push(WINDOWS_MANIFEST_FILE);
+
+        println!("cargo:rerun-if-changed={WINDOWS_MANIFEST_FILE}");
+        // Embed the Windows application manifest file.
+        println!("cargo:rustc-link-arg-bin=cargo=/MANIFEST:EMBED");
+        println!(
+            "cargo:rustc-link-arg-bin=cargo=/MANIFESTINPUT:{}",
+            manifest.to_str().unwrap()
+        );
+        // Turn linker warnings into errors.
+        println!("cargo:rustc-link-arg-bin=cargo=/WX");
```

Comment:
Warnings as errors outside of CI doesn't sound like a good idea. New warnings might get introduced in the future.

Reply:
I am not familiar with Windows in terms of how often it gets new linker warnings. If it turns out too annoying we can always change.

Reply:
The trouble here is that this might make versions of Cargo harder to build in the future. If a new version of MSVC gets released that throws a warning here, we'll need to patch all old versions of Cargo to even make them build. See https://embeddedartistry.com/blog/2017/05/22/werror-is-not-your-friend/ ("-Werror Introduces a Toolchain Version Dependency"). It's better to only enable `/WX` in CI, this way you can always change it, but don't leave users who want to compile Cargo with errors.

Reply:
Assumed "we" referring to someone needing to build olders Cargo on their own. Although we only support the latest stable Cargo, we should try our best to play nice with downstream packagers. Could you open an issue and the team can discuss there?